src/lib.rs

```diff
@@ -256,6 +256,9 @@ impl State {
         };
         self.get_slot_mut(slot_id)?.drop_session(handle);
         self.sessionmap.remove(&handle);
+        if !self.has_sessions(slot_id)? {
+            self.get_token_from_slot_mut(slot_id)?.logout();
+        }
```

Comment:
You need to add a call to logout also in drop_all_sessions_slot() I think. Otherwise calling C_CloseAllSessions will leave the token logged in.

Probably we should add a test for that one too.

Reply:
Note that it may make sense to put this into a helper function to be called at the end of fn_close_session() and fn_close_all_sessions() instead of in drop_session(), as that will make the policy behavior more evident.

Reply:
I consider this more like a specification than a policy that should need to be evident from reading the code in the lib.rs. Especially in case of CloseAllSessions, we do not need to check if there are some leftover sessions, making the logout unconditional.
For the CloseAllSessions, I put the logout directly into lib though as this was the place where I was able to easily access the token. In the CloseSession, I would have to pull much more stuff (slot_id, slot) to check there are some other sessions, which was already available in the `drop_session()` and would duplicate a lot of code in there.